### PR TITLE
Task/bws 126 add run function functionality

### PR DESCRIPTION
**Pull request by junojei**
localhost_3000_slc_configurate-function_a6fcf86d-a154-429c-a9bc-d8f1f7b616c3
localhost_3000_slc_configurate-function_a6fcf86d-a154-429c-a9bc-d8f1f7b616c3 (1)

### CODE REVIEW

File: backend/src/services/aws/lambda/lambda.service.ts [hunk 1]
```diff
@@ -93,6 +100,23 @@ class Lambda {
...
   }
+
+  public async runFunction(name: string): Promise<string> {
+    const { Payload, StatusCode } = await this.#lambdaClient.send(
+      new InvokeCommand({
+        FunctionName: name,
+      }),
+    );
+
+    if (StatusCode !== 200 || !Payload) {
```
Comment (what1s1ove) [inline backend/src/services/aws/lambda/lambda.service.ts]:
```suggestion
    if (StatusCode !== HttpCode.SUCCESS || !hasPayload) {
```

File: frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx [hunk 2]
```diff
@@ -8,17 +8,19 @@ import {
   useState,
 } from 'hooks/hooks';
 import { SLCFunctionConfigurate as SLCFunctionActions } from 'store/actions';
+// import { getResponse } from '..//../helpers/helpers';
```
Comment (what1s1ove) [inline frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx]:
```suggestion
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Comment by junojei** [comment]
localhost_3000_slc_configurate-function_533bbe87-f2f5-42ed-9eb8-f1bc95f9ce6a

**Comment by junojei** [comment]
localhost_3000_slc_configurate-function_a6fcf86d-a154-429c-a9bc-d8f1f7b616c3 (2)

File: backend/src/common/enums/exception/exception-message.enum.ts [hunk 3]
```diff
@@ -25,6 +26,7 @@ const ExceptionMessage = {
   FUNCTION_NOT_FOUND: 'Function not found.',
   FUNCTION_NOT_CHANGE: 'Function code no changes.',
   FUNCTION_NOT_UPDATED: 'Function code not updated.',
+  FUNCTION_NOT_RUN: 'Function not run.',
```
Comment (Yaroslaww-1) [inline backend/src/common/enums/exception/exception-message.enum.ts]:
```suggestion
  FUNCTION_NOT_RUN: 'An error occurred on function run .',
```

File: frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx [hunk 4]
```diff
@@ -45,43 +51,53 @@ const EditForm: FC<Props> = ({ id }) => {
...
+            btnStyle={ButtonStyle.FILLED}
+            label="Run"
+            onClick={handleRun}
+          />
+          <Button
+            className={styles.button}
+            btnStyle={ButtonStyle.FILLED}
+            label="Save"
+            onClick={handleSaveCode}
+            disabled={isPossibleSave ? false : true}
```
Comment (what1s1ove) [inline frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx]:
```suggestion
            disabled={!isPossibleSave}
```

File: frontend/package.json [hunk 5]
```diff
@@ -32,7 +30,8 @@
     "react-router-dom": "6.2.1",
     "react-scripts": "5.0.0",
     "react-select": "5.2.2",
-    "react-table": "7.7.0"
+    "react-table": "7.7.0",
+    "underscore": "1.13.2"
```
Comment (what1s1ove) [inline frontend/package.json]: remove this package and write a function by your self

File: frontend/package.json [hunk 6]
```diff
@@ -41,6 +40,7 @@
     "@types/react-dom": "17.0.11",
     "@types/react-redux-toastr": "7.6.2",
     "@types/react-table": "7.7.9",
+    "@types/underscore": "1.11.4",
```
Comment (what1s1ove) [inline frontend/package.json]: remove this package and write a function by your self

File: frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx [hunk 7]
```diff
@@ -8,25 +8,27 @@ import {
   useState,
 } from 'hooks/hooks';
 import { SLCFunctionConfigurate as SLCFunctionActions } from 'store/actions';
+import { debounce } from 'underscore';
```
Comment (what1s1ove) [inline frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx]:
```suggestion
```

File: frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx [hunk 8]
```diff
@@ -45,43 +51,53 @@ const EditForm: FC<Props> = ({ id }) => {
     );
   };
 
+  const handleOnChangeValue = debounce(setEditCode, 500);
```
Comment (what1s1ove) [inline frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx]: add constant for timeout

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/services/aws/lambda/lambda.service.ts [hunk 9]
```diff
@@ -93,6 +100,28 @@ class Lambda {
...
+      }),
+    );
+    const hasPayload = Boolean(Payload);
+
+    if (StatusCode !== HttpCode.OK || !hasPayload) {
+      throw new SLCError({
+        status: HttpCode.INTERNAL_SERVER_ERROR,
+        message: ExceptionMessage.FUNCTION_NOT_RUN,
+      });
+    }
```
Comment (Yaroslaww-1) [inline backend/src/services/aws/lambda/lambda.service.ts]: Two things here that are not quite correct me:
1. INTERNAL_SERVER_ERROR as a status. In 99% of cases there is no need to **explicitly** throw 500 error. 500 code is ok for an unhandled exceptions. But here you have explicitly defined catch
2. message is obscure. There are a lof of reasons for function to fail on run. Imagine you are using lambda service as a utility tool without knowing how it works internally. And than you receive FUNCTION_NOT_RUN exception. How would you debug it?)

Additionally, it seems like status code does not represent error. For example, everyone from [this thread](https://github.com/aws/aws-sdk-js-v3/issues/2007) use try/catch.
If you still want to provide more detailed exception you can use try/catch in s3.service createBucket method as an example

**Review by Yaroslaww-1 — CHANGES_REQUESTED** [review]

File: frontend/src/components/common/button/button.tsx [hunk 10]
```diff
@@ -23,6 +23,7 @@ type Props = {
   btnStyle?: ButtonStyle;
   btnColor?: ButtonColor;
   className?: string;
+  disabled?: boolean;
```
Comment (what1s1ove) [inline frontend/src/components/common/button/button.tsx]:
```suggestion
  isDisabled?: boolean;
```

File: frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx [hunk 11]
```diff
@@ -8,25 +8,28 @@ import {
...
     }),
   );
   const [editCode, setEditCode] = useState<string>('');
   const dispatch = useAppDispatch();
   const hasSave = loadFunction?.sourceCode === editCode;
   const hasCodeLoading = dataStatus === DataStatus.PENDING;
-  const isPossibleRun = hasSave && !hasCodeLoading;
   const isPossibleSave = !hasSave && !hasCodeLoading;
+  const hasResponse = Boolean(response);
+  const TIMEOUT = 500;
```
Comment (what1s1ove) [inline frontend/src/components/slc-configurate-function/components/edit-form/edit-form.tsx]: *B16*

File: frontend/src/helpers/timeout/timeout.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+const debounce = <A = unknown, R = void>(
```
Comment (what1s1ove) [inline frontend/src/helpers/timeout/timeout.ts]: move it to the shared folder

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: shared/src/helpers/timeout/timeout.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
+const debounce = <A = unknown, R = void>(
```
Comment (what1s1ove) [inline shared/src/helpers/timeout/timeout.ts]: move to `shared/src/helpers/timeout/debounce.helper.ts` and then do reexport

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by Yaroslaww-1 — APPROVED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**